sycl/source/detail/queue_impl.cpp

```diff
@@ -415,6 +415,120 @@ queue_impl::submit_impl(const detail::type_erased_cgfo_ty &CGF,
 }
 #endif
 
+#ifdef __INTEL_PREVIEW_BREAKING_CHANGES
+
+std::vector<ArgDesc> queue_impl::extractArgsAndReqsFromLambda(
+    char *LambdaPtr, detail::kernel_param_desc_t (*ParamDescGetter)(int),
+    size_t NumKernelParams) {
+
+  size_t IndexShift = 0;
+  std::vector<ArgDesc> Args;
+
+  Args.reserve(NumKernelParams);
+
+  for (size_t I = 0; I < NumKernelParams; ++I) {
+    detail::kernel_param_desc_t ParamDesc = ParamDescGetter(I);
+    void *Ptr = LambdaPtr + ParamDesc.offset;
+    const detail::kernel_param_kind_t &Kind = ParamDesc.kind;
+    const int &Size = ParamDesc.info;
+
+    Args.emplace_back(Kind, Ptr, Size, I + IndexShift);
+  }
+
+  return Args;
+}
+
+detail::EventImplPtr queue_impl::submit_direct_impl(
+    const NDRDescT &NDRDesc, const v1::SubmissionInfo &SubmitInfo,
+    const v1::KernelRuntimeInfo &KRInfo, bool CallerNeedsEvent,
+    const detail::code_location &CodeLoc, bool IsTopCodeLoc) {
+  (void)SubmitInfo;
+
+  std::unique_ptr<detail::CG> CommandGroup;
+  detail::CG::StorageInitHelper CGData;
+  std::vector<detail::ArgDesc> Args;
+  std::vector<std::shared_ptr<detail::stream_impl>> StreamStorage;
+  std::vector<std::shared_ptr<const void>> AuxiliaryResources;
+
+  std::unique_lock<std::mutex> Lock(MMutex);
+
+  // Graphs are not supported yet for the no-handler path
+  assert(!hasCommandGraph());
+
+  // No special captures supported yet for the no-handler path
+  assert(!KRInfo.KernelHasSpecialCaptures());
+
+  // Set the No Last Event Mode to false, since the no-handler path
+  // does not support it yet.
+  MNoLastEventMode.store(false, std::memory_order_relaxed);
+
+  // Used by queue_empty() and getLastEvent()
+  MEmpty.store(false, std::memory_order_release);
+
+  // Sync with an external event
+  std::optional<event> ExternalEvent = popExternalEvent();
+  if (ExternalEvent) {
+    CGData.MEvents.push_back(getSyclObjImpl(*ExternalEvent));
+  }
+
+  // Sync with the last event for in order queue
+  EventImplPtr &LastEvent = MDefaultGraphDeps.LastEventPtr;
+  if (isInOrder() && LastEvent) {
+    CGData.MEvents.push_back(LastEvent);
+  }
+
+  // Barrier and un-enqueued commands synchronization for out or order queue
+  if (!isInOrder()) {
+    MMissedCleanupRequests.unset(
+        [&](MissedCleanupRequestsType &MissedCleanupRequests) {
+          for (auto &UpdatedGraph : MissedCleanupRequests)
+            doUnenqueuedCommandCleanup(UpdatedGraph);
+          MissedCleanupRequests.clear();
+        });
+
+    if (MDefaultGraphDeps.LastBarrier &&
+        !MDefaultGraphDeps.LastBarrier->isEnqueued()) {
+      CGData.MEvents.push_back(MDefaultGraphDeps.LastBarrier);
+    }
+  }
+
+  Args = extractArgsAndReqsFromLambda(KRInfo.GetKernelFuncPtr(),
```

Comment:
In the case of the path with a handler, are we extracting args on every submission? Can we cache it somehow?